instrumentation/tomcat-7.0/javaagent/src/main/java/io/opentelemetry/javaagent/instrumentation/tomcat7/ServerHandlerAdvice.java

```diff
@@ -0,0 +1,94 @@
+
+import io.opentelemetry.context.Context;
+import io.opentelemetry.context.Scope;
+import io.opentelemetry.instrumentation.api.servlet.AppServerBridge;
+import io.opentelemetry.javaagent.instrumentation.servlet.v3_0.TagSettingAsyncListener;
+import java.util.concurrent.atomic.AtomicBoolean;
+import net.bytebuddy.asm.Advice;
+import org.apache.coyote.Adapter;
+import org.apache.coyote.Request;
+import org.apache.coyote.Response;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+@SuppressWarnings("unused")
+public class ServerHandlerAdvice {
+  private static final Logger log = LoggerFactory.getLogger(ServerHandlerAdvice.class);
+
+  @Advice.OnMethodEnter(suppress = Throwable.class)
+  public static void onEnter(
+      @Advice.This Adapter itsMe,
+      @Advice.Argument(0) Request request,
+      @Advice.Local("otelContext") Context context,
+      @Advice.Local("otelScope") Scope scope) {
+    Context attachedContext = tracer().getServerContext(request);
+    if (attachedContext != null) {
+      log.debug("Unexpected context found before server handler even started: {}", attachedContext);
+      return;
+    }
+
+    context = tracer().startServerSpan(request, itsMe.getClass().getSimpleName());
+
+    scope = context.makeCurrent();
+  }
+
+  @Advice.OnMethodExit(onThrowable = Throwable.class, suppress = Throwable.class)
+  public static void stopSpan(
+      @Advice.Argument(0) Request request,
+      @Advice.Argument(1) Response response,
+      @Advice.Thrown Throwable throwable,
+      @Advice.Local("otelContext") Context context,
+      @Advice.Local("otelScope") Scope scope) {
+    if (scope == null) {
+      return;
+    }
+    scope.close();
+
+    if (context == null) {
+      return;
+    }
+
+    Throwable throwableToReport =
+        throwable != null ? throwable : AppServerBridge.getUnhandledThrowable(context);
+
+    if (throwableToReport != null) {
+      if (response.isCommitted()) {
+        tracer().endExceptionally(context, throwableToReport, response);
+      } else {
+        tracer().endExceptionally(context, throwableToReport);
+      }
+      return;
+    }
+
+    if (response.isCommitted()) {
+      tracer().end(context, response);
+      return;
+    }
+
+    Object note = request.getNote(1);
+    if (note instanceof org.apache.catalina.connector.Request) {
+      AtomicBoolean responseHandled = new AtomicBoolean(false);
+
+      org.apache.catalina.connector.Request servletRequest =
+          (org.apache.catalina.connector.Request) note;
+      if (servletRequest.isAsync()) {
+        try {
+          servletRequest
+              .getAsyncContext()
+              .addListener(new TagSettingAsyncListener(responseHandled, context));
+        } catch (IllegalStateException e) {
+          // thrown by tomcat if request was already handled while setting the listener.
+          tracer().end(context, response);
+        }
+      }
+    }
```

Comment:
Jetty and Servlet advice also does this, do you think applicable here also?
```suggestion

      // Check again in case the request finished before adding the listener.
      if (!request.isAsync() && responseHandled.compareAndSet(false, true)) {
        tracer().end(context, response);
      }
    }
```

Reply:
@vovencij ?

Reply:
Checked Tomcat's source code, the check is relevant for older tomcats (7). Newer tomcats (e.g., 9) will throw IllegalStateException in this case, which is handled above. Added the check.